File: android/core/protocol/src/main/kotlin/app/hapi/protocol/chat/MessageEnvelope.kt
package app.hapi.protocol.chat

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

/**
 * Envelope helpers ported from `shared/src/messages.ts` (the subset the chat
 * pipeline consumes). `DecryptedMessage.content` is a role-wrapped envelope;
 * see `docs/api/client-contract/messages.md` ("Envelope").
 */

/** `{ role, content, meta? }` — the unwrapped record. [meta] is verbatim wire. */
data class RoleWrappedRecord(
    val role: String,
    /** Present by contract (`'content' in value`); JSON null stays [kotlinx.serialization.json.JsonNull]. */
    val content: JsonElement,
    val meta: JsonElement?,
)

private fun asRoleWrappedRecord(value: JsonElement?): RoleWrappedRecord? {
    val record = asObject(value) ?: return null
    val role = asString(record["role"]) ?: return null
    if (!record.containsKey("content")) return null
    return RoleWrappedRecord(role = role, content = record.getValue("content"), meta = record["meta"])
}

/**
 * `unwrapRoleWrappedRecordEnvelope`: the value itself, or probe
 * `value.message`, `value.data.message`, `value.payload.message` in order.
 */
fun unwrapRoleWrappedRecordEnvelope(value: JsonElement?): RoleWrappedRecord? {
    asRoleWrappedRecord(value)?.let { return it }
    val record = asObject(value) ?: return null

    asRoleWrappedRecord(record["message"])?.let { return it }
    asRoleWrappedRecord(asObject(record["data"])?.get("message"))?.let { return it }
    asRoleWrappedRecord(asObject(record["payload"])?.get("message"))?.let { return it }
    return null
}

private val VISIBLE_CLAUDE_SYSTEM_SUBTYPES = setOf(
    "api_error",
    "turn_duration",
    "microcompact_boundary",
    "compact_boundary",
    "away_summary",
)

fun isClaudeChatVisibleSystemSubtype(subtype: JsonElement?): Boolean =
    asString(subtype) in VISIBLE_CLAUDE_SYSTEM_SUBTYPES

/** `isClaudeChatVisibleMessage` over the raw `data` record's type/subtype. */
fun isClaudeChatVisibleMessage(type: JsonElement?, subtype: JsonElement?): Boolean {
    val typeString = asString(type)
    if (typeString == "rate_limit_event") return false
    if (typeString == "tool_progress") return false
    if (typeString != "system") return true
    return isClaudeChatVisibleSystemSubtype(subtype)
}

private val GOAL_STATUS_REGEX =
    Regex("^Goal (active|paused|complete|blocked|limited by (?:budget|usage))(?:$|\\s+·\\s+)")

/** `isRedundantGoalStatusMessageText` — hub-side goal status echoes. */
fun isRedundantGoalStatusMessageText(value: String?): Boolean {
    if (value == null) return false
    val message = value.trim()
    return message == "Goal cleared" || GOAL_STATUS_REGEX.containsMatchIn(message)
}

/** JSON-object contract shared by decrypted content: an object with a string `type`. */
internal fun typedRecordOrNull(value: JsonElement?): Pair<JsonObject, String>? {
    val record = asObject(value) ?: return null
    val type = asString(record["type"]) ?: return null
    return record to type
}

File: ios/Packages/HapiKit/Sources/HapiProtocol/Chat/MessageEnvelope.swift
import Foundation

// Port of the envelope helpers the chat pipeline uses from
// shared/src/messages.ts: role-wrapped envelope unwrapping, the Claude
// chat-visibility filter, and the redundant goal-status text test.

/// Port of `RoleWrappedRecord`.
public struct RoleWrappedRecord: Equatable, Sendable {
    public var role: String
    public var content: JSONValue
    public var meta: JSONValue?

    public init(role: String, content: JSONValue, meta: JSONValue? = nil) {
        self.role = role
        self.content = content
        self.meta = meta
    }
}

private func asRoleWrappedRecord(_ value: JSONValue?) -> RoleWrappedRecord? {
    guard let object = value?.objectValue else { return nil }
    guard let role = object["role"]?.stringValue else { return nil }
    // `'content' in value` — key presence, any value (JSON knows no undefined).
    guard let content = object["content"] else { return nil }
    return RoleWrappedRecord(role: role, content: content, meta: object["meta"])
}

/// Port of `unwrapRoleWrappedRecordEnvelope`: the value itself, or the
/// wrapped record probed at `message`, `data.message`, `payload.message`.
public func unwrapRoleWrappedRecordEnvelope(_ value: JSONValue) -> RoleWrappedRecord? {
    if let record = asRoleWrappedRecord(value) { return record }
    guard value.objectValue != nil else { return nil }

    if let record = asRoleWrappedRecord(value["message"]) { return record }
    if let record = asRoleWrappedRecord(value["data"]?["message"]) { return record }
    if let record = asRoleWrappedRecord(value["payload"]?["message"]) { return record }
    return nil
}

/// Port of `VISIBLE_CLAUDE_SYSTEM_SUBTYPES`.
private let visibleClaudeSystemSubtypes: Set<String> = [
    "api_error",
    "turn_duration",
    "microcompact_boundary",
    "compact_boundary",
    "away_summary",
]

/// Port of `isClaudeChatVisibleSystemSubtype`.
public func isClaudeChatVisibleSystemSubtype(_ subtype: JSONValue?) -> Bool {
    guard let subtype = subtype?.stringValue else { return false }
    return visibleClaudeSystemSubtypes.contains(subtype)
}

/// Port of `isClaudeChatVisibleMessage` — `type`/`subtype` straight off the
/// wire record (non-string `type` values compare unequal, like in JS).
public func isClaudeChatVisibleMessage(type: JSONValue?, subtype: JSONValue?) -> Bool {
    if type == .string("rate_limit_event") { return false }
    if type == .string("tool_progress") { return false }
    if type != .string("system") { return true }
    return isClaudeChatVisibleSystemSubtype(subtype)
}

/// Port of `isRedundantGoalStatusMessageText`. `·` is U+00B7.
public func isRedundantGoalStatusMessageText(_ value: String?) -> Bool {
    guard let value else { return false }
    let message = value.jsTrimmed
    if message == "Goal cleared" { return true }
    return goalStatusRegex.test(message)
}

private let goalStatusRegex = JSRegex(
    "\\AGoal (active|paused|complete|blocked|limited by (?:budget|usage))(?:\\z|\\s+\u{00B7}\\s+)"
)
